### PR TITLE
Optimize VMUL/VADD part with AVX intrinsic function

### PR DESCRIPTION
**Pull request by yihuaxu**
Used AVX register as the intermediate to save the delay during reading/writing memory frequently.

Platform: Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz / Intel(R) Xeon(R) CPU E5-2699 v3 @ 2.30GHz
Model Path: 
                Text Classification: third_party/inference_demo/text_classification/model
                RNN1: third_party/inference_demo/rnn1/model
Batch Size: 1
Frame Size: Text Classification 128 / RNN1 15
Command: 
                  Text Classification: cd build && ./paddle/fluid/inference/tests/api/test_analyzer_text_classification --infer_model=third_party/inference_demo/text_classification/model --infer_data=third_party/inference_demo/text_classification/data.txt --paddle_num_threads=1 --repeat=1 --batch_size=1 --test_all_data --num_threads=1
                  RNN1: cd build && ./paddle/fluid/inference/tests/api/test_analyzer_rnn1 --infer_model=third_party/inference_demo/rnn1/model --infer_data=third_party/inference_demo/rnn1/data.txt --paddle_num_threads=1 --repeat=1000 --batch_size=1 --test_all_data --num_threads=1 --profiler=1
Data Source: 
                Text Classification: build/third_party/inference_demo/text_classification/data.txt
                RNN: build/third_party/inference_demo/rnn1/data.txt

The following is the comparison with the different scenarios.

image

### CODE REVIEW

**Comment by yihuaxu** [comment]
@tensor-tang Please help me review this commit to implement VMUL and VADD replacement.

**Comment by tensor-tang** [comment]
直接替换需要比较完整的测试下不同模型的性能和在AVX2机器上的性能，比如RNN1的提升如何？另外文本分类的frame size是多大？

**Comment by yihuaxu** [comment]
已经更新好测试状态了

File: paddle/fluid/operators/fusion_lstm_op.cc [hunk 1]
```diff
@@ -270,6 +270,204 @@ class FuisonLSTMKernel : public framework::OpKernel<T> {
...
+    int i = 0;                                     \
+                                                   \
+    __m256 tmp0, tmp1;                             \
+    for (i = 0; i < end; i += block) {             \
+      GET_CT_MOVE_ONE_STEP(ct_1, gates, ct);       \
+    }                                              \
+                                                   \
+    if (rest > 0) {                                \
+      i = D - 8;                                   \
+      GET_CT_MOVE_ONE_STEP(ct_1, gates, ct);       \
```
Comment (tpatejko) [inline paddle/fluid/operators/fusion_lstm_op.cc:299]: I guess, this part is a loop remainder. Shouldn't this part be done sequentially? If `rest` is lower than `block`, you might read some invalid data that follow the buffer you're reading from with `_mm256_loadu_ps` intrinsics.

File: paddle/fluid/operators/fusion_lstm_op.cc [hunk 2]
```diff
@@ -270,6 +270,204 @@ class FuisonLSTMKernel : public framework::OpKernel<T> {
...
+  _mm256_storeu_ps(reinterpret_cast<float*>(ct) + i, tmp0)
+
+// gates: W_ch, W_ih, W_fh, W_oh
+#define GET_Ct(ct_1, gates, ct)                    \
+  /* C_t = C_t-1 * fgated + cand_gated * igated*/  \
+  act_cand(D, gates, gates);                       \
+  if ((D >= 8) && std::is_same<float, T>::value) { \
+    constexpr int block = 8;                       \
+    const int rest = D % block;                    \
+    const int end = D - rest;                      \
```
Comment (tpatejko) [inline paddle/fluid/operators/fusion_lstm_op.cc:289]: I think `rest` and `end` could be marked `constexpr`.

**Comment by yihuaxu** [comment]
This is backup scheme for LSTM VMUL/VADD optimization. Now close it.

**Pull request closed without merging**